### PR TITLE
Rest service

### PR DESCRIPTION
**Pull request by fewi**
Basic REST service according to usage specification.

### CODE REVIEW

**Comment by jayeff** [comment]
Thanks for your PR. I will review it within the next few days.

File: src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package com.inxmail.clicktracker.controller
+
+import com.inxmail.clicktracker.entity.PostUrl
+import com.inxmail.clicktracker.entity.Url
```
Comment (jayeff) [inline src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy]: Can you explain your reasoning to use two separate entities? What's the difference between `PostUrl` and `Url`?

File: src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestMethod
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.bind.annotation.ResponseBody
+
+/**
+ * Created by fewi on 18.06.2015.
+ */
```
Comment (jayeff) [inline src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy]: Please remove these comments on all files. Version control will remember just fine who is the original author of a file.

If you are not familiar with the book Clean Code by Uncle Bob I would highly recommend it to you. We have a copy in the library. Chapter 4 about Comments is interesting in this context.

File: src/main/groovy/com/inxmail/clicktracker/entity/Url.groovy [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.inxmail.clicktracker.entity
+
+import com.fasterxml.jackson.annotation.JsonFormat
```
Comment (jayeff) [inline src/main/groovy/com/inxmail/clicktracker/entity/Url.groovy:3]: Interesting that a company named FasterXML provides _the_ Java JSON library. :sunglasses:

File: src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+@Controller
+class ApiController {
```
Comment (jayeff) [inline src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy]: I would suggest to rename this to UrlController. API is IMHO more the namespace while the Controller should probably reflect the Entity.

**Comment by jayeff** [comment]
I'm missing tests. I suspect that the code was not written in a test driven way?!? :smile:

If you are interested we could pair on re-writing the same code with TDD.

File: src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy (hunk 1 above)
Comment (fewi) [inline src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy]: Yes you are right. We could use Url entity also for POST, dont know what I thought.

File: src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy (hunk 2 above)
Comment (fewi) [inline src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy]: Default settings of my IDE ;)

File: src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy (hunk 4 above)
Comment (fewi) [inline src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy]: OK

**Comment by fewi** [comment]
Unfortunately I haven't written the code in TDD. :)
Yes I'm interested in pair programming for TDD :+1:

File: src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy (hunk 2 above)
Comment (jayeff) [inline src/main/groovy/com/inxmail/clicktracker/controller/ApiController.groovy]: That's what I thought. :wink: Maybe change the setting if you haven't already.

**Comment by jayeff** [comment]
@fewi Let's find an hour next week to get started.

**Comment by jayeff** [comment]
PR #6 was merged, which basically does the some stuff.

**Pull request closed without merging**